File: mindspore/python/mindspore/ops/_op_impl/tbe/reduce_std.py
# Copyright 2022 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

"""ReduceStd op"""
from mindspore.ops.op_info_register import op_info_register, TBERegOp, DataType

reduce_std_op_info = TBERegOp("ReduceStd") \
    .fusion_type("OPAQUE") \
    .async_flag(False) \
    .binfile_name("reduce_std.so") \
    .compute_cost(10) \
    .kernel_name("reduce_std") \
    .partial_flag(True) \
    .attr("axis", "optional", "listInt", "all", "()") \
    .attr("unbiased", "optional", "bool", "all", "true") \
    .attr("keep_dims", "optional", "bool", "all", "false") \
    .input(0, "input_x", False, "required", "all") \
    .output(0, "output_std", False, "required", "all") \
    .output(1, "output_mean", False, "required", "all") \
    .dtype_format(DataType.F16_NCHW, DataType.F16_NCHW, DataType.F16_NCHW) \
    .dtype_format(DataType.F16_NHWC, DataType.F16_NHWC, DataType.F16_NHWC) \
    .dtype_format(DataType.F16_Default, DataType.F16_Default, DataType.F16_Default) \
    .dtype_format(DataType.F32_NCHW, DataType.F32_NCHW, DataType.F32_NCHW) \
    .dtype_format(DataType.F32_NHWC, DataType.F32_NHWC, DataType.F32_NHWC) \
    .dtype_format(DataType.F32_Default, DataType.F32_Default, DataType.F32_Default) \
    .get_op_info()


@op_info_register(reduce_std_op_info)
def _reduce_std_tbe():
    """ReduceStd TBE register"""
    return

File: mindspore/core/ops/ops_func_impl/reduce_std.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ops/ops_func_impl/reduce_std.h"
#include "ops/ops_func_impl/reduce_arithmetic.h"
#include "ops/op_utils.h"
#include "utils/check_convert_utils.h"

namespace mindspore {
namespace ops {
BaseShapePtr ReduceStdFuncImpl::InferShape(const PrimitivePtr &primitive,
                                           const std::vector<AbstractBasePtr> &input_args) const {
  // For ReduceStd, the inputs order: x, axis, unbiased, keep_dims
  // For Reduce base infer function, the inputs order: x, axis, keep_dims
  MS_CHECK_VALUE(input_args.size() == 4, CheckAndConvertUtils::FormatCheckIntegerMsg(
                                           "input_args number", SizeToLong(input_args.size()), kEqual, 4, primitive));
  auto tmp_input_args = std::vector<AbstractBasePtr>{input_args[0], input_args[1], input_args[3]};
  auto shape = ReduceInferShape(primitive, tmp_input_args);
  return std::make_shared<abstract::TupleShape>(std::vector<abstract::BaseShapePtr>(2, shape));
}

TypePtr ReduceStdFuncImpl::InferType(const PrimitivePtr &primitive,
                                     const std::vector<AbstractBasePtr> &input_args) const {
  MS_EXCEPTION_IF_NULL(input_args[0]);
  MS_EXCEPTION_IF_NULL(input_args[0]->GetType());
  return std::make_shared<Tuple>(std::vector<TypePtr>(2, input_args[0]->GetType()->Clone()));
}
}  // namespace ops
}  // namespace mindspore
